### ISSUE TITLE
MME crash unknown cause

### ISSUE THREAD

**Issue by jpad-freedomfi** [magma/magma#4778]
### Your Environment

- **Version:** 1.3.2 with patched MME for Timezone https://github.com/magma/magma/commit/e5638c75a79947f0ce5f23be30bf098810819691 and Paging Timer Expiry 4318
- **Affected Component:** Access Gateway
- **Affected Subcomponent:** MME
- **Deployment Environment:** bare-metal (AGW)

### Describe the Issue

MME crash at customer site. This unit was patched to add swap mem in 2GB deployment of AGW, and patched for mobilityd and eventd memory consumption issues. 

Unclear to us what the source of crash was from logs. It appears mem usage spiked right before the crash, but not clear what service was consuming the memory. 

Event happened at 12:27:35 UTC in logs.

Note the spike in mem usage at this time (image time is UTC -7)
![Screen Shot 2021-02-04 at 12 11 34 PM](https://user-images.githubusercontent.com/61293007/106962624-a0c52b00-66fc-11eb-86ac-f0b172c10f1a.png)

There is no core dump file for this crash. Below is addresssanitizer output
```
Feb  4 12:27:35 agw sctpd[16534]: I0204 12:27:35.410543 16714 sctp_connection.cpp:225] [sd:23] msg of len 39 on 191:1
Feb  4 12:27:35 agw mme[16614]: =================================================================
Feb  4 12:27:35 agw mme[16614]: ==16614==ERROR: AddressSanitizer: unknown-crash on address 0x2039383974000030 at pc 0x7f1f3dfaff7f bp 0x7f1f2e6e7b70 sp 0x7f1f2e6e7320
Feb  4 12:27:35 agw mme[16614]: READ of size 8141 at 0x2039383974000030 thread T5
Feb  4 12:27:35 agw sctpd[16534]: I0204 12:27:35.578017 16714 sctp_connection.cpp:172] HandleClientSock sd = 23
Feb  4 12:27:35 agw sctpd[16534]: I0204 12:27:35.578101 16714 sctp_connection.cpp:225] [sd:23] msg of len 26 on 191:1
Feb  4 12:27:35 agw mme[16614]:     #0 0x7f1f3dfaff7e  (/usr/lib/x86_64-linux-gnu/libasan.so.3+0x5cf7e)
Feb  4 12:27:35 agw mme[16614]:     #1 0x7f1f37f5dd80  (/usr/lib/x86_64-linux-gnu/libzmq.so.5+0x65d80)
Feb  4 12:27:35 agw mme[16614]:     #2 0x7f1f37f4de19  (/usr/lib/x86_64-linux-gnu/libzmq.so.5+0x55e19)
Feb  4 12:27:35 agw mme[16614]:     #3 0x7f1f37f1e1e6  (/usr/lib/x86_64-linux-gnu/libzmq.so.5+0x261e6)
Feb  4 12:27:35 agw mme[16614]:     #4 0x7f1f37f1ca6d  (/usr/lib/x86_64-linux-gnu/libzmq.so.5+0x24a6d)
Feb  4 12:27:35 agw mme[16614]:     #5 0x7f1f37f58a5b  (/usr/lib/x86_64-linux-gnu/libzmq.so.5+0x60a5b)
Feb  4 12:27:35 agw mme[16614]:     #6 0x7f1f3db314a3 in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x74a3)
Feb  4 12:27:35 agw mme[16614]:     #7 0x7f1f39ea5d0e in __clone (/lib/x86_64-linux-gnu/libc.so.6+0xe8d0e)
Feb  4 12:27:35 agw mme[16614]: ==16614==AddressSanitizer CHECK failed: ../../../../src/libsanitizer/asan/asan_report.cc:346 "((0 && "Address is not in memory and not in shadow?")) != (0)" (0x0, 0x0)
Feb  4 12:27:35 agw mme[16614]:     #0 0x7f1f3e01debd  (/usr/lib/x86_64-linux-gnu/libasan.so.3+0xcaebd)
Feb  4 12:27:35 agw mme[16614]:     #1 0x7f1f3e023b13 in __sanitizer::CheckFailed(char const*, int, char const*, unsigned long long, unsigned long long) (/usr/lib/x86_64-linux-gnu/libasan.so.3+0xd0b13)
Feb  4 12:27:35 agw mme[16614]:     #2 0x7f1f3e0191b7  (/usr/lib/x86_64-linux-gnu/libasan.so.3+0xc61b7)
Feb  4 12:27:35 agw mme[16614]:     #3 0x7f1f3e01a118  (/usr/lib/x86_64-linux-gnu/libasan.so.3+0xc7118)
Feb  4 12:27:35 agw mme[16614]:     #4 0x7f1f3e01d376  (/usr/lib/x86_64-linux-gnu/libasan.so.3+0xca376)
Feb  4 12:27:35 agw mme[16614]:     #5 0x7f1f3dfaff9d  (/usr/lib/x86_64-linux-gnu/libasan.so.3+0x5cf9d)
Feb  4 12:27:35 agw mme[16614]:     #6 0x7f1f37f5dd80  (/usr/lib/x86_64-linux-gnu/libzmq.so.5+0x65d80)
Feb  4 12:27:35 agw mme[16614]:     #7 0x7f1f37f4de19  (/usr/lib/x86_64-linux-gnu/libzmq.so.5+0x55e19)
Feb  4 12:27:35 agw mme[16614]:     #8 0x7f1f37f1e1e6  (/usr/lib/x86_64-linux-gnu/libzmq.so.5+0x261e6)
Feb  4 12:27:35 agw mme[16614]:     #9 0x7f1f37f1ca6d  (/usr/lib/x86_64-linux-gnu/libzmq.so.5+0x24a6d)
Feb  4 12:27:35 agw mme[16614]:     #10 0x7f1f37f58a5b  (/usr/lib/x86_64-linux-gnu/libzmq.so.5+0x60a5b)
Feb  4 12:27:35 agw mme[16614]:     #11 0x7f1f3db314a3 in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x74a3)
Feb  4 12:27:35 agw mme[16614]:     #12 0x7f1f39ea5d0e in __clone (/lib/x86_64-linux-gnu/libc.so.6+0xe8d0e)
Feb  4 12:27:35 agw systemd[1]: magma@mme.service: Main process exited, code=exited, status=1/FAILURE

```

**Additional context**
Link to logs files (MME log is big)
MME log
syslog
messages
https://drive.google.com/file/d/1K67Yni7WCtriODbm3dWm0kRdnHlrjIHm/view?usp=sharing

**Comment by berezovskyi-oleksandr** [comment]
@sourabh-nanoti @ulaskozat 
Today we got a crash that is similar to this one (backtrace from gdb):
```
Core was generated by `/usr/local/bin/mme -c /var/opt/magma/tmp/mme.conf -s /var/opt/magma/tmp/spgw.co'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  __memmove_sse2_unaligned_erms () at ../sysdeps/x86_64/multiarch/../multiarch/memmove-vec-unaligned-erms.S:363
363	../sysdeps/x86_64/multiarch/../multiarch/memmove-vec-unaligned-erms.S: No such file or directory.
[Current thread is 1 (Thread 0x7f02a56e9700 (LWP 26513))]
(gdb) bt
#0  __memmove_sse2_unaligned_erms () at ../sysdeps/x86_64/multiarch/../multiarch/memmove-vec-unaligned-erms.S:363
#1  0x00007f02b500ebc1 in __interceptor_memcpy (to=0x625000232933, from=0x7f021d000015, size=8141)
    at ../../../../src/libsanitizer/asan/asan_interceptors.cc:436
#2  0x00007f02aefbcd81 in memcpy (__len=<optimized out>, __src=<optimized out>, __dest=<optimized out>)
    at /usr/include/x86_64-linux-gnu/bits/string3.h:53
#3  zmq::encoder_base_t<zmq::v2_encoder_t>::encode (this=0x608000044520, data_=0x7f02a56e7bc0, size_=<optimized out>)
    at src/encoder.hpp:127
#4  0x00007f02aeface1a in zmq::stream_engine_t::out_event (this=0x61a00002a080) at src/stream_engine.cpp:400
#5  0x00007f02aef7d1e7 in zmq::io_thread_t::in_event (this=0x6110000241c0) at src/io_thread.cpp:85
#6  0x00007f02aef7ba6e in zmq::epoll_t::loop (this=0x611000024080) at src/epoll.cpp:188
#7  0x00007f02aefb7a5c in thread_routine (arg_=0x611000024100) at src/thread.cpp:100
#8  0x00007f02b4b904a4 in start_thread (arg=0x7f02a56e9700) at pthread_create.c:456
#9  0x00007f02b0f04d0f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:97
```

**Comment by jpad-freedomfi** [comment]
@sourabh-nanoti FYI this is an AccessParks site. this site seems to reproduce this issue. seems like frequency is increasing. so far looks like we have Feb 4th, feb 8th, now feb 10th.

**Comment by jpad-freedomfi** [comment]
looks like this is known issue
https://github.com/zeromq/libzmq/issues/4111

**Comment by amarpad** [comment]
The zmq issue is a derivative of something I think. Do you have some MME logs to see what signalling was happening then?

**Comment by jpad-freedomfi** [comment]
Here are logs from Feb8th crash. 
https://drive.google.com/file/d/1F8KL0Ofuu7TuyaY7wuc2xH22Gehqh2Ch/view?usp=sharing 

I will collect logs from today's crash and post shortly.

**Comment by jpad-freedomfi** [comment]
logs from Feb10th crash 
https://drive.google.com/file/d/1hxSui28_2lM1i2TgO56dMn1DomHuicCH/view?usp=sharing 

one odd thing is the amount of white space in the MME log. I think unrelated, but this is the site we have the SSD filling up with logs. Haven't had a chance to fully restart services in the hopes of reducing the log spamming.

**Comment by amarpad** [comment]
Do you have the paging fix? Just making sure its not the same issue

**Comment by jpad-freedomfi** [comment]
yes, we have patched this MME with #4318. Is that the one you meant?

**Comment by talkhasib** [comment]
Would you please attach the MME binary with cores

**Comment by jpad-freedomfi** [comment]
[1.3.2_mme_tz_plus_4318.zip](https://github.com/magma/magma/files/5973573/1.3.2_mme_tz_plus_4318.zip)
@talkhasib

**Comment by talkhasib** [comment]
Thanks
________________________________

**Comment by talkhasib** [comment]
Trying gdb with the provided binary and cores does not seem to match. Can you confirm @jpad-freedomfi

**Comment by jpad-freedomfi** [comment]
whoa, ok. I was sure. But I was wrong. Here is the real one. Which @amarpad means that this particular site does not in fact have the 4318 paging expiry fix. 
[1.3.2.mme_tz_fix_only.zip](https://github.com/magma/magma/files/5974968/1.3.2.mme_tz_fix_only.zip)

**Comment by amarpad** [comment]
@jpad-freedomfi can you upgrade this box? We don't have a better way to debug this issue, so the paging fix would help.

**Comment by jpad-freedomfi** [comment]
Yes. Mentioned the need for an upgrade to customer. Will do it one night soon.

**Comment by jpad-freedomfi** [comment]
@amarpad the customer site has been updated to include #4318 patched MME. Had a somewhat lucky window tonight.

**Comment by jpad-freedomfi** [comment]
@amarpad update here. We have 2 sites still triggering the libzmq crash of mme. One site (Canopy Oaks) has been stable for >1week, but had a MME crash today that looks related to libzmq (logs attached). 

[canopy_oaks_feb25th_libzmq_mme_crash.log](https://github.com/magma/magma/files/6053265/canopy_oaks_feb25th_libzmq_mme_crash.log)

We also have a 2nd site that appears to be producing the crash almost daily. Seems like a good chance to instrument a site to learn more if that is possible.

**Comment by amarpad** [comment]
Can you also upload the syslog? when the MME exists it prints out the logs of memory that gets leaked.
Also, can you check if there is a core?

**Comment by jpad-freedomfi** [comment]
There is not a core file on either machine in /tmp/
here is syslog (too big to post directly) snippet is posted above. crash is at 06:31:26

https://drive.google.com/file/d/1BtzVdUuoh-8UJ9OvokhzYHCy5AY3ScT1/view?usp=sharing

**Comment by amarpad** [comment]
ok this is a pipelined crash
Feb 25 06:30:38 agw pipelined[25872]: Traceback (most recent call last):
Feb 25 06:30:38 agw pipelined[25872]:   File "/usr/lib/python3.5/runpy.py", line 193, in _run_module_as_main
Feb 25 06:30:38 agw pipelined[25872]:     "__main__", mod_spec)
Feb 25 06:30:38 agw pipelined[25872]:   File "/usr/lib/python3.5/runpy.py", line 85, in _run_code
Feb 25 06:30:38 agw pipelined[25872]:     exec(code, run_globals)
Feb 25 06:30:38 agw pipelined[25872]:   File "/usr/local/lib/python3.5/dist-packages/magma/pipelined/main.py", line 157, in <module>
Feb 25 06:30:38 agw pipelined[25872]:     main()
Feb 25 06:30:38 agw pipelined[25872]:   File "/usr/local/lib/python3.5/dist-packages/magma/pipelined/main.py", line 141, in main
Feb 25 06:30:38 agw pipelined[25872]:     service.run()
Feb 25 06:30:38 agw pipelined[25872]:   File "/usr/local/lib/python3.5/dist-packages/magma/common/service.py", line 234, in run
Feb 25 06:30:38 agw pipelined[25872]:     self._loop.run_forever()
Feb 25 06:30:38 agw pipelined[25872]:   File "/usr/lib/python3/dist-packages/aioeventlet.py", line 253, in run_forever
Feb 25 06:30:38 agw pipelined[25872]:     super(EventLoop, self).run_forever()
Feb 25 06:30:38 agw pipelined[25872]:   File "/usr/lib/python3.5/asyncio/base_events.py", line 421, in run_forever
Feb 25 06:30:38 agw pipelined[25872]:     self._run_once()
Feb 25 06:30:38 agw pipelined[25872]:   File "/usr/lib/python3.5/asyncio/base_events.py", line 1388, in _run_once
Feb 25 06:30:38 agw pipelined[25872]:     event_list = self._selector.select(timeout)
Feb 25 06:30:38 agw pipelined[25872]:   File "/usr/lib/python3/dist-packages/aioeventlet.py", line 167, in select
Feb 25 06:30:38 agw pipelined[25872]:     events = self._read_events()
Feb 25 06:30:38 agw pipelined[25872]:   File "/usr/lib/python3/dist-packages/aioeventlet.py", line 162, in _read_events
Feb 25 06:30:38 agw pipelined[25872]:     key = self.get_key(fd)
Feb 25 06:30:38 agw pipelined[25872]:   File "/usr/lib/python3.5/selectors.py", line 191, in get_key
Feb 25 06:30:38 agw pipelined[25872]:     raise KeyError("{!r} is not registered".format(fileobj)) from None
Feb 25 06:30:38 agw pipelined[25872]: KeyError: '19 is not registered'


Joey are you ok if we create a new issue for this and leave this to track the MME issue? @wallyrb

**Comment by jpad-freedomfi** [comment]
Yes, definitely. However best to track the issue. 
Here is another log with MME libzmq crash. 
timestamp Feb 21 05:27:02
https://drive.google.com/file/d/1q_N7ZGwaz3HkP7KSrM8g0h0-vlWEjecA/view?usp=sharing

**Comment by ulaskozat** [comment]
> Yes, definitely. However best to track the issue.
> Here is another log with MME libzmq crash.
> timestamp Feb 21 05:27:02
> https://drive.google.com/file/d/1q_N7ZGwaz3HkP7KSrM8g0h0-vlWEjecA/view?usp=sharing

@jpad-freedomfi this does not look like a segfault:
`Feb 21 05:27:03 agw systemd[1]: magma@mme.service: Main process exited, code=exited, status=1/FAILURE`

**Comment by jpad-freedomfi** [comment]
@ulaskozat Ok, I guess I'm unclear on why MME restarted on Feb 21. Are you saying this portion of the syslog doesn't indicate a problem? 

```
Feb 21 05:27:02 agw mme[5536]: =================================================================
Feb 21 05:27:02 agw mme[5536]: ==5536==ERROR: AddressSanitizer: unknown-crash on address 0x00063980002a at pc 0x7f2de99b2f7f bp 0x7f2dda0e7b70 sp 0x7f2dda0e7320
Feb 21 05:27:02 agw mme[5536]: READ of size 8141 at 0x00063980002a thread T5
Feb 21 05:27:03 agw mme[5536]:     #0 0x7f2de99b2f7e  (/usr/lib/x86_64-linux-gnu/libasan.so.3+0x5cf7e)
Feb 21 05:27:03 agw mme[5536]:     #1 0x7f2de3960d80  (/usr/lib/x86_64-linux-gnu/libzmq.so.5+0x65d80)
Feb 21 05:27:03 agw mme[5536]:     #2 0x7f2de3950e19  (/usr/lib/x86_64-linux-gnu/libzmq.so.5+0x55e19)
Feb 21 05:27:03 agw mme[5536]:     #3 0x7f2de39211e6  (/usr/lib/x86_64-linux-gnu/libzmq.so.5+0x261e6)
Feb 21 05:27:03 agw mme[5536]:     #4 0x7f2de391fa6d  (/usr/lib/x86_64-linux-gnu/libzmq.so.5+0x24a6d)
Feb 21 05:27:03 agw mme[5536]:     #5 0x7f2de395ba5b  (/usr/lib/x86_64-linux-gnu/libzmq.so.5+0x60a5b)
Feb 21 05:27:03 agw mme[5536]:     #6 0x7f2de95344a3 in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x74a3)
Feb 21 05:27:03 agw mme[5536]:     #7 0x7f2de58a8d0e in __clone (/lib/x86_64-linux-gnu/libc.so.6+0xe8d0e)
Feb 21 05:27:03 agw mme[5536]: Address 0x00063980002a is located in the shadow gap area.
Feb 21 05:27:03 agw mme[5536]: SUMMARY: AddressSanitizer: unknown-crash (/usr/lib/x86_64-linux-gnu/libasan.so.3+0x5cf7e)
Feb 21 05:27:03 agw mme[5536]: Thread T5 created by T0 here:
Feb 21 05:27:03 agw mme[5536]:     #0 0x7f2de9986f59 in __interceptor_pthread_create (/usr/lib/x86_64-linux-gnu/libasan.so.3+0x30f59)
Feb 21 05:27:03 agw mme[5536]:     #1 0x7f2de395bb72  (/usr/lib/x86_64-linux-gnu/libzmq.so.5+0x60b72)
Feb 21 05:27:03 agw mme[5536]: ==5536==ABORTING
Feb 21 05:27:03 agw systemd[1]: magma@mme.service: Main process exited, code=exited, status=1/FAILURE

```

**Comment by amarpad** [comment]
Yeah I think Ulas was looking at the restart that corresponded to pipelined restarting. This is a legit issue.

**Comment by talkhasib** [comment]
Looked at the log. Not easy to see much without a core. Wonder why a core is not getting generated? @ulaskozat

**Comment by jpad-freedomfi** [comment]
We built this MME binary based on 1.3.2 tag zip file and two patches per the issue. Are there any build flags or other patches I can apply to force more helpful logging or core files?

**Comment by amarpad** [comment]
@talkhasib this is an asan restart due to an invalid memory access there is no core by default for this.

**Comment by talkhasib** [comment]
Found few issues with misaligned memory access that might be the cause of the problem.
https://github.com/magma/magma/pull/5262

Once the PR lands, I will cherry pick into 1.3 and 1.4 and ping you

**Comment by jpad-freedomfi** [comment]
Ninja.

**Comment by talkhasib** [comment]
@jpad-freedomfi

We backported the following two fixes to v1.3 branch
https://github.com/magma/magma/pull/5262
https://github.com/magma/magma/pull/5307

We believe these might help with the ZMQ issue.

To enable ASAN and UBSAN on ZMQ, we need to build LIBZMQ and CZMQ from source with ASAN enabled
I managed to do that on my local setup. Not statically linked though

**Comment by jpad-freedomfi** [comment]
per Tariq's majorly awesome patch above, this can be closed.
